File: server/src/battle/model/shootable/ShootableObject.java
package battle.model.shootable;

import battle.model.Coordinate;
import model.battle.map.Position;
import util.battle.BattleUtils;

public class ShootableObject {

    public int id;
    public double baseHp;
    public double currentHP;
    public boolean isDied = false;
    public double hitRadius;
    public Position position = new Position(0, 0);

    protected ShootableObject(int id) {
        this.id = id;
    }

    public int getId(){

        return this.id;
    }

    public void decreaseHP(double amount){

        setCurrentHP(Math.max(0, this.currentHP - amount));
        if (this.currentHP == 0) {
            this.isDied = true;
        }
    }

    public void decreaseHpByBullet(double amount){

        this.decreaseHP(amount);
    }

    public void increaseHP(double amount){

        setCurrentHP(Math.min(this.baseHp, this.currentHP + amount));
    }

    public void setCurrentHP(double currentHP) {

        this.currentHP = BattleUtils.round(currentHP);
    }

    public Coordinate getPosition() {
        return new Coordinate(this.position.getX(), this.position.getY());
    }

    public boolean canTarget() {
        return true;
    }

}

File: client/src/modules/base/shootable/ShootableObject.js
let ShootableObject = cc.Class.extend({

    ctor: function (id){
        this.id = id;
        this.isDied = false;
        this.baseHp = 0;
        this.currentHP = 0;
        this.hitRadius = 0;
        this.position = cc.p(0, 0);
    },
    getId: function () {
        return this.id;
    },
    decreaseHP: function (amount) {
        this.setCurrentHP(Math.max(0, this.currentHP - amount))
        if (this.currentHP == 0) {
            this.isDied = true
        }
    },
    increaseHP: function (amount) {
        if(amount > 0 && this.currentHP < this.baseHp){
            this.buffHpUpAni = true;
        }
        this.setCurrentHP(Math.min(this.baseHp, this.currentHP + amount))
    },
    setCurrentHP: function (hp) {
        this.currentHP = Utils.round(hp)
    },

    getPosition: function () {
        return cc.p(this.position);
    },

    canTarget: function () {
        return true;
    }

})